### PR TITLE
Added freeCompilationResources method to compiledFunction

### PR DESCRIPTION
**Pull request by gcatron**
Summary:
Added freeCompilationResources method to compiledFunction
 and added implementation of it for OpenCLFunction. 
This method allows for freeing resources in a compiledFunction that are no longer needed after the function has been loaded on a device.
Added pass in provisioner to call new method

Documentation: Updated Backends.md to include new method.

Test Plan: ninja check

### CODE REVIEW

File: lib/Backends/OpenCL/OpenCL.cpp [hunk 1]
```diff
@@ -124,6 +124,10 @@ OpenCLFunction::OpenCLFunction(std::unique_ptr<IRFunction> F,
   setTraceInfo(std::move(traceInfo));
 }
 
+void OpenCLFunction::freeCompilationResources() {
+  runtimeBundle_.freeConstants();
```
Comment (beicy) [inline lib/Backends/OpenCL/OpenCL.cpp]: Can I ask why only constants need to be freed?

Comment (gcatron) [reply lib/Backends/OpenCL/OpenCL.cpp]: They are copied for initialization of the device after which they are no longer needed so they are freed. The symbolTable is still needed.

File: include/glow/Runtime/Provisioner/Provisioner.h [hunk 2]
```diff
@@ -64,8 +64,8 @@ class Provisioner final {
   /// List of available DeviceManagers added during initialization.
   std::vector<DeviceManager *> devices_;
 
-  /// Helper function to cleanup a failed provision call.
-  void cleanupProvision(llvm::ArrayRef<std::string> names);
+  /// Helper function to cleanup a provision call.
+  void cleanupProvision(llvm::ArrayRef<std::string> names, bool failure = true);
```
Comment (beicy) [inline include/glow/Runtime/Provisioner/Provisioner.h]: Please add a few words about \p failure.

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 3]
```diff
@@ -239,9 +239,17 @@ llvm::Error Provisioner::removeFunction(llvm::StringRef name) {
...
 
-void Provisioner::cleanupProvision(llvm::ArrayRef<std::string> names) {
+void Provisioner::cleanupProvision(llvm::ArrayRef<std::string> names,
+                                   bool failure) {
   std::lock_guard<std::mutex> functionLock(functionsLock_);
   for (auto &name : names) {
     activeFunctions_.erase(name);
+    if (failure) {
+      // Remove any functions added before the failure.
+      functions_.erase(name);
```
Comment (beicy) [inline lib/Runtime/Provisioner/Provisioner.cpp]: here do we need to free the resource as well?

**Pull request closed without merging**